lib/port_scanner.py

```diff
@@ -0,0 +1,45 @@
+import queue
+import socket
+from threading import Thread, Lock
+from queue import Queue
+
+N_THREADS = 150
+
+queue = Queue()
+print_lock = Lock()
+
+def scan_port(host, port):
+    try:
+        sock = socket.socket()
+        sock.connect((host, port))
+    except:
+        pass
+    else:
+        with print_lock:
+            print(f"{host:15}:{port:5} is open")
+    finally:
+        sock.close()
+
+
+def scan_thread(host):
+    global queue
+    while True:
+        port_number = queue.get()
```

Comment:
Port scanning is a perfectly parallelizable task. You just split the range into multiple ranges with math, and spawn a thread, each with a smaller range to check. Eg. 256 ports into 4 threads would give 4 ranges: `0-63`, `64-127`, `128-191`, `192-256`. This is not only simpler to implement (just a little bit of math) but also way more performant than queues. (not that performance is our main concert right now but yeah)

Reply:
I'll do this improvement later today because it is 1am